template-manager-cmtemplate/src/main/java/com/sequenceiq/cloudbreak/cmtemplate/servicetype/ServiceTypeResolver.java

```diff
@@ -0,0 +1,68 @@
+package com.sequenceiq.cloudbreak.cmtemplate.servicetype;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+import org.apache.commons.collections.CollectionUtils;
+import org.apache.commons.lang3.StringUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Component;
+
+import com.sequenceiq.cloudbreak.cmtemplate.CmTemplateProcessor;
+import com.sequenceiq.cloudbreak.exception.BadRequestException;
+
+@Component
+public class ServiceTypeResolver {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ServiceTypeResolver.class);
+
+    private final List<ServiceTypeMatcher> serviceTypeMatchers;
+
+    public ServiceTypeResolver(List<ServiceTypeMatcher> serviceTypeMatchers) {
+        this.serviceTypeMatchers = serviceTypeMatchers;
+    }
+
+    /**
+     * Run all service type providers, in case of none of them will be applied, the default DATAHUB service type will be used.
+     * If it will match multiple serviceTypes, that will fall back the default DATAHUB service type.
+     * If there is a desired service type (provided by an another application) and the result won't be the same that will cause a runtime exception.
+     * @param cmTemplateProcessor template processor that contains role and configuration data
+     * @param desiredServiceType desired type service type that is provided by other service
+     * @return resolved service type for the template
+     */
+    public String resolveServiceType(CmTemplateProcessor cmTemplateProcessor, String desiredServiceType) {
+        Set<ServiceType> serviceTypes = new HashSet<>();
+        ServiceType result = ServiceType.DATAHUB;
+        if (StringUtils.isNotBlank(desiredServiceType) && !ServiceType.isValueValid(desiredServiceType)) {
+            String errorMessage = String.format("'%s' is not a valid value for ServiceType enum.", desiredServiceType);
+            throw new BadRequestException(errorMessage);
+        }
+        if (CollectionUtils.isNotEmpty(serviceTypeMatchers)) {
+            for (ServiceTypeMatcher serviceTypeMatcher : serviceTypeMatchers) {
+                if (serviceTypeMatcher.match(cmTemplateProcessor)) {
+                    serviceTypes.add(serviceTypeMatcher.getServiceType());
+                }
+            }
+        }
+        serviceTypes.remove(ServiceType.DATAHUB);
+        if (serviceTypes.size() == 0) {
+            LOGGER.debug("Service type only matched for dummy/default DataHub provider. Default DATAHUB service type will be used.");
+        } else if (serviceTypes.size() == 1) {
+            result = serviceTypes.iterator().next();
+            LOGGER.debug("The following service type will be selected based on the service type providers: {}", result.value());
+        } else {
+            LOGGER.warn("You cannot match on multiple service types based on the service type providers. " +
```

Comment:
maybe in case the user could not specify the servicetype then we should check the stack.type. What do you think?

Reply:
default service type is DATAHUB here, is the same as the stack type (that is not used for DATALAKE). metering v1 event supports only that. better to use that default enum (in case of DATAHUB name will change in the future or something), so i think it should be fine to use the default ServiceType enum instead of the stack type